src/DataMapper/SubscriptionMapper.php

```diff
@@ -0,0 +1,122 @@
+        }
+        $this->connection
+            ->insert(
+                'subscription',
+                [
+                    'activity' => $data->activity instanceof \stdClass ? $data->activity->id : null,
+                    'visitor' => $data->visitor->id,
+                    'type' => $data->type,
+                    'time_from' => $data->startDay->date,
+                    'period' => $data->period,
+                    'available' => $data->status === 'Available' ? 1 : 0
+                ]
+            );
+        $subscription->persists((int)$this->connection->lastInsertId());
+    }
+
+    /**
+     * @param int $id
+     * @param VisitorMapper $visitorMapper
+     * @param ActivityMapper $activityMapper
+     * @return SubscriptionInterface
+     * @throws LoadFromDbImpossible
+     * @throws \Doctrine\DBAL\Exception
+     */
+    public function find(
+        int $id,
+        VisitorMapper $visitorMapper,
+        ActivityMapper $activityMapper
+    ): SubscriptionInterface {
+
+        $data = $this->connection->fetchAssociative('SELECT * FROM subscription WHERE id=?', [$id]);
+        $subscriptionType = $data['type'];
+        $subscriptionTime = new \DateTimeImmutable($data['time_from']);
+        $subscriptionPeriod = getDateInterval((int)$data['period']);
+        $subscriptionVisitor = $visitorMapper->find((int)$data['visitor']);
+
+        switch ($subscriptionType) {
+            case 'GroupLife\Core\Subscription\Activity':
+                $newSubscription = new $subscriptionType(
+                    $subscriptionTime,
+                    $subscriptionPeriod,
+                    $activityMapper->find(
+                        (int)$data['activity'],
+                        new LeaderMapper($this->connection),
+                        new ScheduleMapper($this->connection)
+                    ),
+                    $subscriptionVisitor
+                );
+                $newSubscription->persists($id);
+                break;
+            case 'GroupLife\Core\Subscription\Membership':
+                $newSubscription = new $subscriptionType(
+                    $subscriptionTime,
+                    $subscriptionPeriod,
+                    $subscriptionVisitor
+                );
+                break;
+            case 'GroupLife\Core\Subscription\FixedTime':
+                $newSubscription = new $subscriptionType(
+                    $subscriptionTime,
+                    $subscriptionVisitor
+                );
+                break;
+            case 'GroupLife\Core\Subscription\OneTime':
+                $newSubscription = new $subscriptionType(
+                    $subscriptionTime,
+                    $subscriptionPeriod,
+                    $subscriptionVisitor,
+                    $data['available'] === '1' ? 'Available' : 'Not available'
+                );
+                $newSubscription->persists($id);
```

Comment:
Don't you need to notify any subscription about its ID? It is messed for FixedTime and Membership.

Reply:
Fixed